File: common/src/main/java/org/tdf/common/trie/AbstractTrie.java
package org.tdf.common.trie;

import lombok.NonNull;
import org.tdf.common.serialize.Codec;
import org.tdf.common.store.Store;
import org.tdf.common.util.HexBytes;

import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Function;

abstract class AbstractTrie<K, V> implements Trie<K, V> {
    abstract Codec<K> getKCodec();

    abstract Codec<V> getVCodec();

    public abstract Store<byte[], byte[]> getStore();

    abstract V getFromBytes(byte[] data);

    abstract void putBytes(byte[] key, byte[] value);

    abstract void removeBytes(byte[] data);

    @Override
    public void set(@NonNull K k, @NonNull V val) {
        putBytes(getKCodec().getEncoder().apply(k), getVCodec().getEncoder().apply(val));
    }

    @Override
    public V get(@NonNull K k) {
        byte[] data = getKCodec().getEncoder().apply(k);
        return getFromBytes(data);
    }

    @Override
    public void remove(@NonNull K k) {
        byte[] data = getKCodec().getEncoder().apply(k);
        removeBytes(data);
    }

    public void traverse(BiFunction<? super K, ? super V, Boolean> traverser) {
        traverseInternal((k, v) -> traverser.apply(getKCodec().getDecoder().apply(k), getVCodec().getDecoder().apply(v)));
    }

    public void traverseValue(Function<? super V, Boolean> traverser) {
        traverseInternal((k, v) -> traverser.apply(getVCodec().getDecoder().apply(v)));
    }

    abstract void traverseInternal(BiFunction<byte[], byte[], Boolean> traverser);
}

File: common/src/main/java/org/tdf/common/trie/AbstractTrie.kt
package org.tdf.common.trie

import org.tdf.common.serialize.Codec
import org.tdf.common.store.Store
import java.util.function.BiFunction
import java.util.function.Function

abstract class AbstractTrie<K, V> : Trie<K, V> {
    abstract val kCodec: Codec<K>
    abstract val vCodec: Codec<V>
    abstract override val store: Store<ByteArray, ByteArray>

    abstract fun getFromBytes(key: ByteArray): V?

    abstract fun putBytes(key: ByteArray, value: ByteArray)
    abstract fun removeBytes(key: ByteArray)
    override fun set(k: K, v: V) {
        putBytes(kCodec.encoder.apply(k), vCodec.encoder.apply(v))
    }

    override fun get(k: K): V? {
        val data = kCodec.encoder.apply(k)
        return getFromBytes(data)
    }

    override fun remove(k: K) {
        val data = kCodec.encoder.apply(k)
        removeBytes(data)
    }

    override fun traverse(traverser: BiFunction<in K, in V, Boolean>) {
        traverseInternal { k: ByteArray, v: ByteArray ->
            traverser.apply(
                kCodec.decoder.apply(k), vCodec.decoder.apply(v)
            )
        }
    }

    override fun traverseValue(traverser: Function<in V, Boolean>) {
        traverseInternal { _: ByteArray, v: ByteArray ->
            traverser.apply(
                vCodec.decoder.apply(v)
            )
        }
    }

    abstract fun traverseInternal(traverser: BiFunction<ByteArray, ByteArray, Boolean>)
}